File: attendance-backend/app/models/employee.py
import uuid
from typing import Optional
from datetime import datetime

from sqlalchemy import String, ForeignKey, DateTime
from app.models.base import GUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.base import TimestampMixin, SoftDeleteMixin


class Employee(TimestampMixin, SoftDeleteMixin, Base):
    __tablename__ = "employees"

    employee_id: Mapped[str] = mapped_column(String(20), unique=True, nullable=False, index=True)
    user_id: Mapped[uuid.UUID] = mapped_column(
        GUID(),
        ForeignKey("users.id", ondelete="CASCADE"),
        unique=True,
        nullable=False,
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    email: Mapped[str] = mapped_column(String(255), nullable=False)
    phone: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    department: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    designation: Mapped[str] = mapped_column(String(100), nullable=False)
    gender: Mapped[Optional[str]] = mapped_column(String(10), nullable=True)  # male, female, other
    joined_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    avatar_glb_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)

    user = relationship("User", back_populates="employee", lazy="noload")
    face_photos = relationship("FacePhoto", back_populates="employee", lazy="selectin")
    check_logs = relationship("CheckLog", back_populates="employee", lazy="noload")

    def __repr__(self) -> str:
        return f"<Employee(name={self.name}, emp_id={self.employee_id})>"

File: employee_app/lib/domain/entities/employee.dart
import 'package:equatable/equatable.dart';

class Employee extends Equatable {
  final String id;
  final String employeeId;
  final String name;
  final String email;
  final String? phone;
  final String department;
  final String designation;
  final String? gender;
  final DateTime joinedAt;
  final bool isActive;
  final bool faceRegistered;
  final String? profilePhotoUrl;
  final bool isNewUser;

  const Employee({
    required this.id,
    required this.employeeId,
    required this.name,
    required this.email,
    this.phone,
    required this.department,
    required this.designation,
    this.gender,
    required this.joinedAt,
    this.isActive = true,
    this.faceRegistered = false,
    this.profilePhotoUrl,
    this.isNewUser = false,
  });

  Employee copyWith({
    String? id,
    String? employeeId,
    String? name,
    String? email,
    String? phone,
    String? department,
    String? designation,
    String? gender,
    DateTime? joinedAt,
    bool? isActive,
    bool? faceRegistered,
    String? profilePhotoUrl,
    bool? isNewUser,
  }) {
    return Employee(
      id: id ?? this.id,
      employeeId: employeeId ?? this.employeeId,
      name: name ?? this.name,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      department: department ?? this.department,
      designation: designation ?? this.designation,
      gender: gender ?? this.gender,
      joinedAt: joinedAt ?? this.joinedAt,
      isActive: isActive ?? this.isActive,
      faceRegistered: faceRegistered ?? this.faceRegistered,
      profilePhotoUrl: profilePhotoUrl ?? this.profilePhotoUrl,
      isNewUser: isNewUser ?? this.isNewUser,
    );
  }

  @override
  List<Object?> get props => [id, employeeId, name, email];
}
